### ISSUE TITLE
calculates the 2-norm partial gradient as coefficient for loss 

### ISSUE THREAD

**Issue by jiasenlu** [google/jax#12755]
I am trying to replicate the VQ-GAN in JAX and FLAX, it works great, but I met an issue with how to calculate the 2-norm partial gradient as the coefficient for loss. The function can be seen here (https://github.com/CompVis/taming-transformers/blob/24268930bf1dce879235a7fddd0b2355b84d7ea6/taming/modules/losses/vqperceptual.py#L63) 

```
    def calculate_adaptive_weight(self, nll_loss, g_loss, last_layer=None):
        if last_layer is not None:
            nll_grads = torch.autograd.grad(nll_loss, last_layer, retain_graph=True)[0]
            g_grads = torch.autograd.grad(g_loss, last_layer, retain_graph=True)[0]
        else:
            nll_grads = torch.autograd.grad(nll_loss, self.last_layer[0], retain_graph=True)[0]
            g_grads = torch.autograd.grad(g_loss, self.last_layer[0], retain_graph=True)[0]

        d_weight = torch.norm(nll_grads) / (torch.norm(g_grads) + 1e-4)
        d_weight = torch.clamp(d_weight, 0.0, 1e4).detach()
        d_weight = d_weight * self.discriminator_weight
        return d_weight
```
The d_weight fraction is used as the weight coefficient of the discriminator to weight the total_loss.
And It calculates the 2-norm ratio after deriving the parameters of the last layer of the model based on rec_loss and g_loss. 

Do I need to nest the grad function inside the loss function? Is there any faster implementation? Thanks

Please:

- [ x] Check for duplicate requests.
- [ x] Describe your goal, and if possible provide a code snippet with a motivating example.
